hack/verify-container-images.sh

```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+TRIVY_VERSION=0.34.0
+
+GO_OS="$(go env GOOS)"
+if [[ "${GO_OS}" == "linux" ]]; then
+  TRIVY_OS="Linux"
+elif [[ "${GO_OS}" == "darwin"* ]]; then
+  TRIVY_OS="macOS"
+fi
+
+GO_ARCH="$(go env GOARCH)"
+if [[ "${GO_ARCH}" == "amd" ]]; then
+  TRIVY_ARCH="32bit"
+elif [[ "${GO_ARCH}" == "amd64"* ]]; then
+  TRIVY_ARCH="64bit"
+elif [[ "${GO_ARCH}" == "arm" ]]; then
+  TRIVY_ARCH="ARM"
+elif [[ "${GO_ARCH}" == "arm64" ]]; then
+  TRIVY_ARCH="ARM64"
+fi
+
+TOOL_BIN=hack/tools/bin
+
+# Downloads trivy scanner
+curl -L -o ${TOOL_BIN}/trivy.tar.gz \
+    https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_${TRIVY_OS}-${TRIVY_ARCH}.tar.gz \
+
+tar xfO ${TOOL_BIN}/trivy.tar.gz trivy > ${TOOL_BIN}/trivy
+chmod +x ${TOOL_BIN}/trivy
+rm ${TOOL_BIN}/trivy.tar.gz
+
+# Builds all the container images to be scanned and cleans up changes to ./*manager_image_patch.yaml ./*manager_pull_policy.yaml.
+make REGISTRY=gcr.io/k8s-staging-cluster-api PULL_POLICY=IfNotPresent TAG=dev docker-build
+make clean-release-git
+
+# Scan the images
+${TOOL_BIN}/trivy image -q gcr.io/k8s-staging-cluster-api/clusterctl-${GO_ARCH}:dev
```

Comment:
I think for now I would start with trying to fix the vulnerabilities when we find them. I"m not sure what we should do when we find vulnerabilities which:
1. don't have a fix yet
2. require bumps in upstream dependencies (e.g. controller-runtime, k/k)

Would we delay the release indefinitely / who can make the assessment how critical the vulnerability is?

But as I said, I'm fine with starting on a best-effort basis and then we can see how it goes. WIth the GitHub dependency scan we didn't run into big problems as we simply had to bump dependencies.

Reply:
> who can make the assessment how critical the vulnerability is?

That's the trickiest part for me, we can rely on SIG security as a last resort, but some of the work will fall on us and TBH I'm not sure how much work it is and how complex it will be.

> what we should do when we find vulnerabilities which:
>
> - don't have a fix yet
> - require bumps in upstream dependencies (e.g. controller-runtime, k/k)

I think the first step is to document those vulnerabilities in the release note.
Then if we can fix, we will do it as soon as possible, otherwise, if it is not possible to fix e.g. for older branches, we will recommend users upgrade. At least they are informed upfront which is an improvement from the current situation.

TL;DR; it is a green field for us, let's start acting according to our capabilities/best judgment, then we can always iterate. But I also plan to bring up this during the office hours meeting and ask the community to step up.

Reply:
Sounds good.

Overall I don't expect problems. We already had the GitHub security dependency scanner running via dependabot (although only on main) and up until now bumping dependencies / Go was sufficient.